### PR TITLE
EASY-2311: email in agreements

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes EASY-

#### When applied it will
* 

#### Before merge
* [x] merge https://github.com/DANS-KNAW/easy-schema/pull/76
* [x] release/deploy `easy-schema`

#### How should this be manually tested?
Build/deploy on deasy together with  ingest-flow and DANS-KNAW/easy-stage-dataset#145
* [x] submit a deposit, examine emd (as either depositor or datamander)

#### Related pull requests on github
* requires https://github.com/DANS-KNAW/easy-schema/pull/76
  for one of the unit tests and ingest-flow of a manual submit
* required by https://github.com/DANS-KNAW/easy-stage-dataset/pull/145

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 1]
```diff
@@ -113,7 +113,7 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
     DepositDir.get(draftBase, user, id)
   }
 
-  def getUser(user: String): Try[Map[String, Seq[String]]] = authentication.getUser(user)
+  def getUserProperties(user: String): Try[Map[String, Seq[String]]] = authentication.getUser(user)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: I'm not quite happy with the return type here. I know it didn't really change in this PR, but why not do the parsing of the `Map` here or in `authentication.getUser` (latter is even better, IMO)? By returning a `User` object, you don't have to parse it all the way down in `AgreementsXml` and hence that class doesn't need to know about the LDAP specific keywords.

File: src/test/scala/nl.knaw.dans.easy.deposit/TestSupportFixture.scala [hunk 2]
```diff
@@ -53,6 +53,12 @@ trait TestSupportFixture extends FlatSpec with Matchers with Inside with BeforeA
     }
   }
 
+  val userMap: Map[String, Seq[String]] = Map(
+    "uid" -> Seq("user001"),
+    "displayName" -> Seq("fullName"),
+    "mail" -> Seq("info@dans.knaw.nl"),
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/TestSupportFixture.scala]: please use the `does not exist` address instead!

File: src/test/scala/nl.knaw.dans.easy.deposit/docs/DDMSpec.scala [hunk 3]
```diff
@@ -591,7 +591,7 @@ trait DdmBehavior {
 
       assume(triedSchema.isAvailable)
       val result = triedSchema.validate(ddm)
-      if (result.isFailure) // trouble shoot broken tests
+      if (result.isFailure) // show the relevant XML section to trouble shoot a broken test
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/docs/DDMSpec.scala]:
```suggestion
      if (result.isFailure) // show the relevant XML section to troubleshoot a broken test
```

File: src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala [hunk 4]
```diff
@@ -43,8 +43,8 @@ class IntegrationSpec extends TestSupportFixture with ServletFixture with Scalat
   private val app: EasyDepositApiApp = new EasyDepositApiApp(minimalAppConfig) {
     override val pidRequester: PidRequester = mockPidRequester
 
-    override def getUser(user: String): Try[Map[String, Seq[String]]] = {
-      Success(Map(("displayName", Seq("F. Bar"))))
+    override def getUserProperties(user: String): Try[Map[String, Seq[String]]] = {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala]: As you do this in both `DepositServletFixture` and here, it begs the question whether you can extract something here... Perhaps `EasyDepositApiAppFixture` (or just `AppFixture`)?

File: src/main/scala/nl.knaw.dans.easy.deposit/docs/AgreementsXml.scala [hunk 5]
```diff
@@ -17,15 +17,16 @@ package nl.knaw.dans.easy.deposit.docs
...
 import scala.xml.Elem
 
 object AgreementsXml extends SchemedXml {
   override val schemaNameSpace = "http://easy.dans.knaw.nl/schemas/bag/metadata/agreements/"
-  override val schemaLocation = "https://easy.dans.knaw.nl/schemas/bag/metadata/agreements/2019/01/agreements.xsd"
+  override val schemaLocation = "https://easy.dans.knaw.nl/schemas/bag/metadata/agreements/2019/09/agreements.xsd"
 
-  def apply(userId: String, dateSubmitted: DateTime, dm: DatasetMetadata, fullname: String): Try[Elem] = {
-    for {
+  def apply(dateSubmitted: DateTime, dm: DatasetMetadata, userProperties: UserInfo): Try[Elem] = {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/docs/AgreementsXml.scala]: How about 
```suggestion
  def apply(dateSubmitted: DateTime, dm: DatasetMetadata, user: UserInfo): Try[Elem] = {
```

change the reference(s) below yourself

File: src/main/scala/nl.knaw.dans.easy.deposit/docs/AgreementsXml.scala [hunk 6]
```diff
@@ -17,15 +17,16 @@ package nl.knaw.dans.easy.deposit.docs
...
+    if (userProperties == null) Failure(new IllegalArgumentException("no user attributes"))
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/docs/AgreementsXml.scala]: `userProperties` CANNOT be `null`! The type is `UserInfo`, not `Option[UserInfo]`.

File: src/main/scala/nl.knaw.dans.easy.deposit/docs/UserInfo.scala [hunk 7]
```diff
@@ -25,24 +26,35 @@ case class UserInfo(userName: String,
...
+          ""
+        }
+    }
     new UserInfo(
 
       // mandatory: https://github.com/DANS-KNAW/dans.easy-ldap-dir/blob/f17c391/files/easy-schema.ldif#L83-L84
-      userName = attributes.getOrElse("uid", Seq.empty).headOption.getOrElse(""),
+      userName = getAttribute("uid"),
 
       firstName = attributes.getOrElse("cn", Seq.empty).headOption,
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/docs/UserInfo.scala]: why not let this use `getAttribute` as well? In other words, why is `firstName` optional?

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 8]
```diff
@@ -113,7 +113,7 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
     DepositDir.get(draftBase, user, id)
   }
 
-  def getUser(user: String): Try[Map[String, Seq[String]]] = authentication.getUser(user)
+  def getUserProperties(user: String): Try[UserInfo] = authentication.getUser(user)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: Let's revert this and call this function `getUser`.

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/UserServlet.scala [hunk 9]
```diff
@@ -26,8 +26,8 @@ import scala.util.Try
 class UserServlet(app: EasyDepositApiApp) extends ProtectedServlet(app) {
 
   get("/") {
-    Try(app.getUser(user.id)).flatten // no throw should slip through
-      .map(properties => Ok(toJson(UserInfo(properties))))
+    Try(app.getUserProperties(user.id)).flatten // no throw should slip through
+      .map(userInfo => Ok(toJson(userInfo)))
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/servlets/UserServlet.scala]:
```suggestion
      .map(user => Ok(toJson(user)))
```

File: src/test/scala/nl.knaw.dans.easy.deposit/SubmitterSpec.scala [hunk 10]
```diff
@@ -57,7 +57,7 @@ class SubmitterSpec extends TestSupportFixture with MockFactory {
       .getOrRecover(e => fail(s"could not get stateManager of test deposit $e"))
     addDoiToDepositProperties(getBag(depositDir))
 
-    createSubmitter(unrelatedGroup).submit(depositDir, stateManager, "fullName", testDir / "staged") should matchPattern {
+    createSubmitter(unrelatedGroup).submit(depositDir, stateManager, UserInfo(userMap), testDir / "staged") should matchPattern {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/SubmitterSpec.scala]: Not sure, but now you're implicitly testing `UserInfo.apply` here as well? Shouldn't there be a `testUser` object around here instead and have separate tests (if they're not there yet) for `UserInfo.apply`?

Same below and in other files...

File: src/test/scala/nl.knaw.dans.easy.deposit/authentication/LdapAuthenticationSpec.scala [hunk 11]
```diff
@@ -111,7 +110,7 @@ class LdapAuthenticationSpec extends TestSupportFixture with MockFactory {
     })
     inside(authentication.getUser("someone")) {
       case Success(user) => // just sampling the result
-        user("dansPrefixes").toArray shouldBe Array("van", "den")
+        user.prefix shouldBe Some("van den")
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/authentication/LdapAuthenticationSpec.scala]:
```suggestion
        user.prefix.value shouldBe "van den"
```

File: src/test/scala/nl.knaw.dans.easy.deposit/docs/AgreementsSpec.scala [hunk 12]
```diff
@@ -27,56 +27,80 @@ class AgreementsSpec extends TestSupportFixture {
...
-      ""
+      userProperties = UserInfo(userMap),
     ) should matchPattern {
       case Failure(e: InvalidDocumentException) if e.getMessage == "invalid DatasetMetadata: Please set PrivacySensitiveDataPresent" =>
     }
   }
 
-  private lazy val triedSchema: Try[Schema] = AgreementsXml.loadSchema
+  it should "log a user without an email" in {
+    // TODO for now set 'trace' at https://github.com/DANS-KNAW/easy-deposit-api/blob/f1b9924/src/test/resources/logback.xml#L15
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/docs/AgreementsSpec.scala]: Is this `TODO` still needed? Same below...

File: src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala [hunk 13]
```diff
@@ -88,7 +83,9 @@ class IntegrationSpec extends TestSupportFixture with ServletFixture with Scalat
       post(
         uri = s"/deposit",
         headers = Seq(fooBarBasicAuthHeader)
-      ) { new String(bodyBytes) }
+      ) {
+        new String(bodyBytes)
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala]: why can't you use `body` here?

File: src/main/scala/nl.knaw.dans.easy.deposit/docs/AgreementsXml.scala (hunk 6 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/docs/AgreementsXml.scala]: https://github.com/DANS-KNAW/easy-deposit-api/blob/d614ef71f78e1d04872892a1a2f795136fea81eb/src/test/scala/nl.knaw.dans.easy.deposit/docs/AgreementsSpec.scala#L86-L88

File: src/test/scala/nl.knaw.dans.easy.deposit/TestSupportFixture.scala [hunk 14]
```diff
@@ -53,6 +54,8 @@ trait TestSupportFixture extends FlatSpec with Matchers with Inside with BeforeA
     }
   }
 
+  val defaultUserInfo = UserInfo("user001", displayName = "fullName", email ="does.not.exist@dans.knaw.nl", lastName="")
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/TestSupportFixture.scala]: formatting!!!

File: src/test/scala/nl.knaw.dans.easy.deposit/authentication/LdapAuthenticationSpec.scala [hunk 15]
```diff
@@ -96,22 +97,45 @@ class LdapAuthenticationSpec extends TestSupportFixture with MockFactory {
...
-    inside(authentication.getUser("someone")) {
-      case Success(user) => // just sampling the result
-        user("dansPrefixes").toArray shouldBe Array("van", "den")
-    }
+    }).getUser("someone")
+    .getOrRecover(e => fail(e))
+    .prefix shouldBe Some("van")
+  }
+
+  // TODO for now set 'trace' at https://github.com/DANS-KNAW/easy-deposit-api/blob/f1b9924/src/test/resources/logback.xml#L15
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/authentication/LdapAuthenticationSpec.scala]: TODO still necessary?

File: src/test/scala/nl.knaw.dans.easy.deposit/authentication/LdapAuthenticationSpec.scala [hunk 16]
```diff
@@ -96,22 +97,45 @@ class LdapAuthenticationSpec extends TestSupportFixture with MockFactory {
...
-    inside(authentication.getUser("someone")) {
-      case Success(user) => // just sampling the result
-        user("dansPrefixes").toArray shouldBe Array("van", "den")
-    }
+    }).getUser("someone")
+      .getOrRecover(e => fail(e))
+      .prefix shouldBe Some("van")
+  }
+
+  // TODO figure out how to intercept and assert the logging
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.deposit/authentication/LdapAuthenticationSpec.scala]: @rvanheest any suggestion?

**Review by rvanheest — APPROVED** [review]

**Pull request merged**